### PR TITLE
Deprecate auto mode from ModelCheckpoint and EarlyStopping

### LINKED ISSUE #4668: ModelCheckpoint automatic mode selection is misleading
## 🐛 Bug

https://github.com/PyTorchLightning/pytorch-lightning/blob/2d78d9b84af96c6fe06a4b2e4a27754a275db460/pytorch_lightning/callbacks/model_checkpoint.py#L299-L305

This behaviour makes a strong assumption that any non-accuracy/fmeasure metric is a loss.

### To Reproduce

I use a metric called `val_pck_0.15`, which stands for Percentage of Correct Keypoints  (popular metric in pose estimation). It was misclassified and time was wasted.

### Expected behavior

There should either be a dictionary with predefined metric names and corresponding configs or IMHO there should be no automatic mode selection, just a MisconfigurationError if mode is not provided.

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #4668 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Deprecated
 
+- Deprecated `mode='auto'` from ModelCheckpoint and EarlyStopping ([#4695](https://github.com/PyTorchLightning/pytorch-lightning/pull/4695))
```
Comment (Borda) [inline CHANGELOG.md:46]:
```suggestion
- Deprecated `mode='auto'` from `ModelCheckpoint` and `EarlyStopping` ([#4695](https://github.com/PyTorchLightning/pytorch-lightning/pull/4695))
```

**Review by Borda — CHANGES_REQUESTED** [review]
we need to update docs

**Review by carmocca — APPROVED** [review]
Nice!

File: pytorch_lightning/callbacks/early_stopping.py [hunk 2]
```diff
@@ -86,21 +94,34 @@ def __init__(self, monitor: str = 'early_stop_on', min_delta: float = 0.0, patie
         # It is set to False initially and overwritten, if eval results have been validated
         self.based_on_eval_results = False
 
-        if mode not in self.mode_dict and mode != 'auto':
+        self.__init_monitor_mode()
+
+        self.min_delta *= 1 if self.monitor_op == torch.gt else -1
```
Comment (Borda) [inline pytorch_lightning/callbacks/early_stopping.py:99]: where is the `self. monitor_op` set?
I can not see it inn init and `self.__init_monitor_mode()` uses `self.mode`

File: pytorch_lightning/callbacks/early_stopping.py [hunk 3]
```diff
@@ -54,7 +54,8 @@ class EarlyStopping(Callback):
             mode it will stop when the quantity
             monitored has stopped increasing; in `auto`
             mode, the direction is automatically inferred
-            from the name of the monitored quantity. Default: ``'auto'``.
+            from the name of the monitored quantity. Default: ``'min'``.
+            Setting ``mode='auto'`` has been deprecated in v1.1 and will be removed in v1.3.
```
Comment (Borda) [inline pytorch_lightning/callbacks/early_stopping.py:58]: can we get this with `.. warning::`

Comment (rohitgr7) [reply pytorch_lightning/callbacks/early_stopping.py:58]: sure.

File: pytorch_lightning/callbacks/early_stopping.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/early_stopping.py:99]: it's a property.
https://github.com/PyTorchLightning/pytorch-lightning/blob/9a5d40aff4bbf8a0d72eae7d8a96e65fdefda234/pytorch_lightning/callbacks/early_stopping.py#L122-L124

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/callbacks/early_stopping.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/callbacks/early_stopping.py:102]: my bad, thx :]

**Review by awaelchli — APPROVED** [review]
very reasonable PR <3 
@rohitgr7 did you do a full text search for mode=auto across the whole project?

File: docs/source/weights_loading.rst [hunk 4]
```diff
@@ -84,10 +84,11 @@ You can also control more advanced options, like `save_top_k`, to save the best
     # saves a file like: my/path/sample-mnist-epoch=02-val_loss=0.32.ckpt
     checkpoint_callback = ModelCheckpoint(
         monitor='val_loss',
-        dirpath='my/path/,
+        dirpath='my/path/',
         filename='sample-mnist-{epoch:02d}-{val_loss:.2f}',
         save_top_k=3,
-        mode='min')
+        mode='min'
```
Comment (Borda) [inline docs/source/weights_loading.rst:90]:
```suggestion
        mode='min',
```

File: tests/test_deprecated.py [hunk 5]
```diff
@@ -106,6 +112,6 @@ def test_end(self, outputs):
         return {'test_loss': torch.tensor(0.7)}
 
 
-def test_auc_reorder_remove_in_v1_1_0():
+def test_reorder_remove_in_v1_1_0():
```
Comment (Borda) [inline tests/test_deprecated.py:115]:
```suggestion
def test_reorder_remove_in_v1_1():
```

**Review by Borda — APPROVED** [review]

**Comment by rohitgr7** [comment]
> @rohitgr7 did you do a full text search for mode=auto across the whole project?

yes @awaelchli

**Comment by rohitgr7** [comment]
@williamFalcon this needs your approval.

**Review by manbldwin — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -144,7 +149,7 @@ def __init__(
         save_last: Optional[bool] = None,
         save_top_k: Optional[int] = None,
         save_weights_only: bool = False,
-        mode: str = "auto",
+        mode: str = "min",
```
Comment (mpariente) [inline pytorch_lightning/callbacks/model_checkpoint.py:152]: If `mode` is set to `"min"` by default in this release, former codebase that didn't register the `mode` (use the default kwargs, `"default"`) won't be notified about the deprecation, and that might cause problems (e.g in case the "auto" mode inferred to use "max").

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: we can keep the auto functionality and during the deprecation phase print the warning that mode will become "min" by default in future.

Comment (mpariente) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: I think that would be a better deprecation process.

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: The drawback is that this warning would show up for every user, even if they don't configure any custom callback. Since the Trainer configures a default checkpoint callback (with mode auto) it could be confusing to users not knowing about the checkpoint callback. 
A way around this could be to have an internal flag `_show_deprecation_warning` on the checkpoint callback and in the Trainer we set it to false when configuring the default one.

Comment (mpariente) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: Or define this particular model checkpoint in the min mode.

Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: how about when it's set to True, we can set the mode to something else and treat it as ``auto`` mode until it gets removed? maybe `default-auto` or `dep-auto` or something else?? and not raise the warning in this case.

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: > The drawback is that this warning would show up for every user, even if they don't configure any custom callback. Since the Trainer configures a default checkpoint callback (with mode auto) it could be confusing to users not knowing about the checkpoint callback.

The default checkpoint callback uses `mode='auto'` but also `monitor=None`. That means it actually defaults to `min` due to this condition:

https://github.com/PyTorchLightning/pytorch-lightning/blob/f677efe61ecf28af1e5216a5baf502be6f394564/pytorch_lightning/callbacks/model_checkpoint.py#L314-L316

So, to keep backwards compatibility and to show a warning only when appropriate, we should keep `mode='auto'` in the ModelCheckpoint signature but `mode='min'` here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/f677efe61ecf28af1e5216a5baf502be6f394564/pytorch_lightning/trainer/connectors/callback_connector.py#L74

This way, users who used a custom ModelCheckpoint without setting the mode would get the warning but users who used the default one wouldn't.

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:152]: Aha, good observation. That makes it much easier:)

**Comment by rohitgr7** [comment]
@williamFalcon mind review?
